src/config/datadog.rs

```diff
@@ -0,0 +1,117 @@
+use super::{ComponentId, Config, SinkOuter, SourceOuter};
+use crate::{
+    sinks::datadog::metrics::DatadogConfig, sources::internal_metrics::InternalMetricsConfig,
+};
+use serde::{Deserialize, Serialize};
+
+static INTERNAL_METRICS_KEY: &str = "#datadog_internal_metrics";
+static DATADOG_METRICS_KEY: &str = "#datadog_metrics";
+
+#[derive(Debug, Deserialize, Serialize, PartialEq, Clone)]
+#[serde(default, deny_unknown_fields)]
+pub struct Options {
+    #[serde(default = "default_enabled")]
+    pub enabled: bool,
+
+    #[serde(default = "default_api_key")]
+    pub api_key: Option<String>,
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        Self {
+            enabled: default_enabled(),
+            api_key: default_api_key(),
+        }
+    }
+}
+
+/// By default, the Datadog feature is enabled.
+fn default_enabled() -> bool {
+    true
+}
+
+/// By default, no API key is provided.
+fn default_api_key() -> Option<String> {
+    None
+}
+
+/// Augment configuration with observability via Datadog if the feature is enabled and
+/// an API key is provided.
+pub fn attach(config: &mut Config) {
+    // Return early if an API key is missing, or the feature isn't enabled.
+    let api_key = match (&config.datadog.api_key, config.datadog.enabled) {
+        (Some(api_key), true) => api_key.clone(),
+        _ => return,
+    };
+
+    info!("Datadog API key detected. Internal metrics will be sent to Datadog.");
```

Comment:
Since `api_key` needs to be explicitly set, I think the user is more likely to be aware at this stage that they've opted in to sending metrics to Datadog.

I do think "detected" here is a misdirection; I could see that being interpreted as we've magically ascertained the value (from their environment, say) rather than it being given by the user.

I'll change that to "provided" to make it clear that the user has explicitly given us their API key.

Reply:
Noting per our recent discussion that I've updated this to require an explicit `enabled = true`, with the API key provided explicitly or (optionally) implicitly by the environment.

I've updated the PR description to reflect.